### DISCUSSION TITLE
code-server install-extension in airgapped environment (ms-toolsai/jupyter extension)

### DISCUSSION THREAD

**Discussion by iszegedi** [coder/code-server#6482] [General] [answered]
Hello,

I am trying to install ms-tools/jupyter extension to my code-server (version 4.16.1) using the downloaded vsix file in an airgapped environment where internet connections are blocked. 

The command I am using is:

extension=""ms-toolsai/jupyter/2023.3.100/file/ms-toolsai.jupyter-2023.3.100.vsix"
code-server --install-extension "./code-server-extensions/${extension}" --extensions-dir ${1}
($1 refers to the code-server extension dir)

The command does not complete, it seems to hang. The same command works on a different machine which is not in airgapped environment and allowed to have internet connection.

I know that jupyter has bundled extensions (ms-toolsai.jupyter-keymap, ms-toolsai.jupyter-renderers, ms-toolsai.vscode-jupyter-slideshow, ms-toolsai.vscode-jupyter-cell-tags ) so I also downloaded those related vsix files to local directory. But it does not help, the extension installation does not complete. By running network monitoring (wireshark) it looks like the code-server --install-extension command is still trying to establish connection to external sites., even if I am using local downloaded vsix files

Any suggestions how I can install this extension in an airgapped envrionment? 

Thanks a lot,
Istvan

**Comment by code-asher** [comment, marked as answer]
Are you installing them all at the same time?  Might have to install the dependencies first, and then Jupyter.  But I think it might still make a query even then to check if they are all the latest version.  If so, this is probably a bug.  Or maybe it is trying to send telemetry?  Where are the requests going?

One workaround is to unzip the visx into the extensions directory instead of using `--install-extension`.

Another is to host your own marketplace somewhere in your local network (https://github.com/coder/code-marketplace) and connect to that.

**Comment by iszegedi** [comment]
I am installing them one by one,  the order of execution is that I install first the four dependencies and only then the mstoolsai/jupyter.  I disabled telemetry and looking for updates, all the suspicious internet communications but it did not help. The request seems to be going to open-vsx.org and also to openvsxorg.blob.core.windows.net

I have tried the work-around with the zip/unzip approach into the extension directory and it seems to be working ok but I would prefer to stick to the "proper"command line option if I can.

**Reply by code-asher** [reply]
Ah OK, I see, thank you for experimenting.  Yeah my guess is that it tries to check to see if it is the latest version.  I am not sure if this is technically a bug but it feels to me like it is, installing a local `.vsix` should not reach out externally IMO.  But I guess it could also be argued that the connection should be immediately rejected instead of hanging which means the bug is in the method of airgapping?

Oh one other thing to try, maybe unsetting the extension gallery will make it stop sending requests.  For example:

```
export EXTENSIONS_GALLERY='{}'
code-server
```

**Reply by iszegedi** [reply]
I tried unsetting the extension gallery variable and it seems to be working fine.

export EXTENSIONS_GALLERY='{}'
code-server --install-extension "EXTENSION NAME"

I am still running a few more tests but it looks good, thanks for the suggestion.
